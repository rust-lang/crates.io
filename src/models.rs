pub use self::action::{NewVersionOwnerAction, VersionAction, VersionOwnerAction};
pub use self::category::{Category, CrateCategory, NewCategory};
pub use self::crate_owner_invitation::{
    CrateOwnerInvitation, NewCrateOwnerInvitation, NewCrateOwnerInvitationOutcome,
};
pub use self::default_versions::{update_default_version, verify_default_version};
pub use self::deleted_crate::NewDeletedCrate;
pub use self::dependency::{Dependency, DependencyKind, ReverseDependency};
pub use self::download::VersionDownload;
pub use self::email::{Email, NewEmail};
pub use self::follow::Follow;
pub use self::keyword::{CrateKeyword, Keyword};
pub use self::krate::{Crate, CrateName, NewCrate, RecentCrateDownloads};
pub use self::owner::{CrateOwner, Owner, OwnerKind};
pub use self::rights::Rights;
pub use self::team::{NewTeam, Team};
pub use self::token::ApiToken;
pub use self::user::{NewUser, User};
pub use self::version::{NewVersion, TopVersions, Version};

pub mod helpers;

mod action;
pub mod category;
pub mod crate_owner_invitation;
pub mod default_versions;
mod deleted_crate;
pub mod dependency;
pub mod download;
mod email;
mod follow;
mod keyword;
pub mod krate;
mod owner;
mod rights;
mod team;
pub mod token;
pub mod user;
pub mod version;
