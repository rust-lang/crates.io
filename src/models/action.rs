use crate::models::{ApiToken, User, Version};
use crate::schema::*;
use bon::Builder;
use chrono::NaiveDateTime;
use crates_io_diesel_helpers::pg_enum;
use diesel::prelude::*;
use diesel_async::{AsyncPgConnection, RunQueryDsl};

pg_enum! {
    pub enum VersionAction {
        Publish = 0,
        Yank = 1,
        Unyank = 2,
    }
}

impl From<VersionAction> for &'static str {
    fn from(action: VersionAction) -> Self {
        match action {
            VersionAction::Publish => "publish",
            VersionAction::Yank => "yank",
            VersionAction::Unyank => "unyank",
        }
    }
}

impl From<VersionAction> for String {
    fn from(action: VersionAction) -> Self {
        let string: &'static str = action.into();

        string.into()
    }
}

#[derive(Debug, Clone, Copy, Queryable, Identifiable, Associations)]
#[diesel(
    table_name = version_owner_actions,
    check_for_backend(diesel::pg::Pg),
    belongs_to(Version),
    belongs_to(User, foreign_key = user_id),
    belongs_to(ApiToken, foreign_key = api_token_id),
)]
pub struct VersionOwnerAction {
    pub id: i32,
    pub version_id: i32,
    pub user_id: i32,
    pub api_token_id: Option<i32>,
    pub action: VersionAction,
    pub time: NaiveDateTime,
}

impl VersionOwnerAction {
    pub async fn all(conn: &mut AsyncPgConnection) -> QueryResult<Vec<Self>> {
        version_owner_actions::table.load(conn).await
    }

    pub async fn by_version(
        conn: &mut AsyncPgConnection,
        version: &Version,
    ) -> QueryResult<Vec<(Self, User)>> {
        use version_owner_actions::dsl::version_id;

        version_owner_actions::table
            .filter(version_id.eq(version.id))
            .inner_join(users::table)
            .order(version_owner_actions::dsl::id)
            .load(conn)
            .await
    }

    pub async fn for_versions(
        conn: &mut AsyncPgConnection,
        versions: &[&Version],
    ) -> QueryResult<Vec<Vec<(Self, User)>>> {
        Ok(Self::belonging_to(versions)
            .inner_join(users::table)
            .order(version_owner_actions::dsl::id)
            .load(conn)
            .await?
            .grouped_by(versions))
    }
}

#[derive(Insertable, Debug, Builder)]
#[diesel(table_name = version_owner_actions, check_for_backend(diesel::pg::Pg))]
pub struct NewVersionOwnerAction {
    version_id: i32,
    user_id: i32,
    api_token_id: Option<i32>,
    #[builder(into)]
    action: VersionAction,
}

impl NewVersionOwnerAction {
    pub async fn insert(&self, conn: &mut AsyncPgConnection) -> QueryResult<VersionOwnerAction> {
        diesel::insert_into(version_owner_actions::table)
            .values(self)
            .get_result(conn)
            .await
    }
}
