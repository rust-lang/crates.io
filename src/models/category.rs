use chrono::NaiveDateTime;
use diesel::{
    delete, dsl, insert_into, sql_query, ExpressionMethods, QueryDsl, QueryResult,
    TextExpressionMethods,
};
use diesel_async::scoped_futures::ScopedFutureExt;
use diesel_async::{AsyncConnection, AsyncPgConnection, RunQueryDsl};
use std::future::Future;

use crate::models::Crate;
use crate::schema::*;

#[derive(Clone, Identifiable, Queryable, QueryableByName, Debug, Selectable)]
#[diesel(table_name = categories, check_for_backend(diesel::pg::Pg))]
pub struct Category {
    pub id: i32,
    pub category: String,
    pub slug: String,
    pub description: String,
    pub crates_cnt: i32,
    pub created_at: NaiveDateTime,
}

type WithSlug<'a> = diesel::dsl::Eq<categories::slug, crate::sql::lower<&'a str>>;

#[derive(Associations, Insertable, Identifiable, Debug, Clone, Copy)]
#[diesel(
    table_name = crates_categories,
    check_for_backend(diesel::pg::Pg),
    primary_key(crate_id, category_id),
    belongs_to(Category),
    belongs_to(Crate),
)]
pub struct CrateCategory {
    crate_id: i32,
    category_id: i32,
}

impl Category {
    pub fn with_slug(slug: &str) -> WithSlug<'_> {
        categories::slug.eq(crate::sql::lower(slug))
    }

    #[dsl::auto_type(no_type_alias)]
    pub fn by_slug<'a>(slug: &'a str) -> _ {
        let filter: WithSlug<'a> = Self::with_slug(slug);
        categories::table.filter(filter)
    }

    pub async fn update_crate(
        conn: &mut AsyncPgConnection,
        crate_id: i32,
        slugs: &[&str],
    ) -> QueryResult<Vec<String>> {
        conn.transaction(|conn| {
            async move {
                let categories: Vec<Category> = categories::table
                    .filter(categories::slug.eq_any(slugs))
                    .load(conn)
                    .await?;

                let invalid_categories = slugs
                    .iter()
                    .filter(|s| !categories.iter().any(|c| c.slug == **s))
                    .map(ToString::to_string)
                    .collect();

                let crate_categories = categories
                    .iter()
                    .map(|c| CrateCategory {
                        category_id: c.id,
                        crate_id,
                    })
                    .collect::<Vec<_>>();

                delete(crates_categories::table)
                    .filter(crates_categories::crate_id.eq(crate_id))
                    .execute(conn)
                    .await?;

                insert_into(crates_categories::table)
                    .values(&crate_categories)
                    .execute(conn)
                    .await?;

                Ok(invalid_categories)
            }
            .scope_boxed()
        })
        .await
    }

    pub async fn count_toplevel(conn: &mut AsyncPgConnection) -> QueryResult<i64> {
        categories::table
            .filter(categories::category.not_like("%::%"))
            .count()
            .get_result(conn)
            .await
    }

    pub fn toplevel(
        conn: &mut AsyncPgConnection,
        sort: &str,
        limit: i64,
        offset: i64,
    ) -> impl Future<Output = QueryResult<Vec<Category>>> {
        use diesel::sql_types::Int8;

        let sort_sql = match sort {
            "crates" => "ORDER BY crates_cnt DESC",
            _ => "ORDER BY category ASC",
        };

        // Collect all the top-level categories and sum up the crates_cnt of
        // the crates in all subcategories
        sql_query(format!(include_str!("toplevel.sql"), sort_sql))
            .bind::<Int8, _>(limit)
            .bind::<Int8, _>(offset)
            .load(conn)
    }

    pub async fn subcategories(&self, conn: &mut AsyncPgConnection) -> QueryResult<Vec<Category>> {
        use diesel::sql_types::Text;

        sql_query(include_str!("../subcategories.sql"))
            .bind::<Text, _>(&self.category)
            .load(conn)
            .await
    }

    /// Gathers the parent categories from the top-level Category to the direct parent of this Category.
    /// Returns categories as a Vector in order of traversal, not including this Category.
    /// The intention is to be able to have slugs or parent categories arrayed in order, to
    /// offer the frontend, for examples, slugs to create links to each parent category in turn.
    pub async fn parent_categories(
        &self,
        conn: &mut AsyncPgConnection,
    ) -> QueryResult<Vec<Category>> {
        use diesel::sql_types::Text;

        sql_query(include_str!("../parent_categories.sql"))
            .bind::<Text, _>(&self.slug)
            .load(conn)
            .await
    }
}

/// Struct for inserting categories; only used in tests. Actual categories are inserted
/// in src/boot/categories.rs.
#[derive(Insertable, AsChangeset, Default, Debug)]
#[diesel(table_name = categories, check_for_backend(diesel::pg::Pg))]
pub struct NewCategory<'a> {
    pub category: &'a str,
    pub slug: &'a str,
    pub description: &'a str,
}

#[cfg(test)]
mod tests {
    use super::*;
    use crates_io_test_db::TestDatabase;
    use diesel_async::RunQueryDsl;

    #[tokio::test]
    async fn category_toplevel_excludes_subcategories() {
        use self::categories;

        let test_db = TestDatabase::new();
        let mut conn = test_db.async_connect().await;

        insert_into(categories::table)
            .values(&vec![
                (
                    categories::category.eq("Cat 2"),
                    categories::slug.eq("cat2"),
                ),
                (
                    categories::category.eq("Cat 1"),
                    categories::slug.eq("cat1"),
                ),
                (
                    categories::category.eq("Cat 1::sub"),
                    categories::slug.eq("cat1::sub"),
                ),
            ])
            .execute(&mut conn)
            .await
            .unwrap();

        let cats = Category::toplevel(&mut conn, "", 10, 0)
            .await
            .unwrap()
            .into_iter()
            .map(|c| c.category)
            .collect::<Vec<_>>();
        let expected = vec!["Cat 1".to_string(), "Cat 2".to_string()];
        assert_eq!(expected, cats);
    }

    #[tokio::test]
    async fn category_toplevel_orders_by_crates_cnt_when_sort_given() {
        use self::categories;

        let new_cat = |category, slug, crates_cnt| {
            (
                categories::category.eq(category),
                categories::slug.eq(slug),
                categories::crates_cnt.eq(crates_cnt),
            )
        };

        let test_db = TestDatabase::new();
        let mut conn = test_db.async_connect().await;

        insert_into(categories::table)
            .values(&vec![
                new_cat("Cat 1", "cat1", 0),
                new_cat("Cat 2", "cat2", 2),
                new_cat("Cat 3", "cat3", 1),
            ])
            .execute(&mut conn)
            .await
            .unwrap();

        let cats = Category::toplevel(&mut conn, "crates", 10, 0)
            .await
            .unwrap()
            .into_iter()
            .map(|c| c.category)
            .collect::<Vec<_>>();
        let expected = vec![
            "Cat 2".to_string(),
            "Cat 3".to_string(),
            "Cat 1".to_string(),
        ];
        assert_eq!(expected, cats);
    }

    #[tokio::test]
    async fn category_toplevel_applies_limit_and_offset() {
        use self::categories;

        let test_db = TestDatabase::new();
        let mut conn = test_db.async_connect().await;

        insert_into(categories::table)
            .values(&vec![
                (
                    categories::category.eq("Cat 1"),
                    categories::slug.eq("cat1"),
                ),
                (
                    categories::category.eq("Cat 2"),
                    categories::slug.eq("cat2"),
                ),
            ])
            .execute(&mut conn)
            .await
            .unwrap();

        let cats = Category::toplevel(&mut conn, "", 1, 0)
            .await
            .unwrap()
            .into_iter()
            .map(|c| c.category)
            .collect::<Vec<_>>();
        let expected = vec!["Cat 1".to_string()];
        assert_eq!(expected, cats);

        let cats = Category::toplevel(&mut conn, "", 1, 1)
            .await
            .unwrap()
            .into_iter()
            .map(|c| c.category)
            .collect::<Vec<_>>();
        let expected = vec!["Cat 2".to_string()];
        assert_eq!(expected, cats);
    }

    #[tokio::test]
    async fn category_toplevel_includes_subcategories_in_crate_cnt() {
        use self::categories;

        let new_cat = |category, slug, crates_cnt| {
            (
                categories::category.eq(category),
                categories::slug.eq(slug),
                categories::crates_cnt.eq(crates_cnt),
            )
        };

        let test_db = TestDatabase::new();
        let mut conn = test_db.async_connect().await;

        insert_into(categories::table)
            .values(&vec![
                new_cat("Cat 1", "cat1", 1),
                new_cat("Cat 1::sub", "cat1::sub", 2),
                new_cat("Cat 2", "cat2", 3),
                new_cat("Cat 2::Sub 1", "cat2::sub1", 4),
                new_cat("Cat 2::Sub 2", "cat2::sub2", 5),
                new_cat("Cat 3", "cat3", 6),
            ])
            .execute(&mut conn)
            .await
            .unwrap();

        let cats = Category::toplevel(&mut conn, "crates", 10, 0)
            .await
            .unwrap()
            .into_iter()
            .map(|c| (c.category, c.crates_cnt))
            .collect::<Vec<_>>();
        let expected = vec![
            ("Cat 2".to_string(), 12),
            ("Cat 3".to_string(), 6),
            ("Cat 1".to_string(), 3),
        ];
        assert_eq!(expected, cats);
    }

    #[tokio::test]
    async fn category_toplevel_applies_limit_and_offset_after_grouping() {
        use self::categories;

        let new_cat = |category, slug, crates_cnt| {
            (
                categories::category.eq(category),
                categories::slug.eq(slug),
                categories::crates_cnt.eq(crates_cnt),
            )
        };

        let test_db = TestDatabase::new();
        let mut conn = test_db.async_connect().await;

        insert_into(categories::table)
            .values(&vec![
                new_cat("Cat 1", "cat1", 1),
                new_cat("Cat 1::sub", "cat1::sub", 2),
                new_cat("Cat 2", "cat2", 3),
                new_cat("Cat 2::Sub 1", "cat2::sub1", 4),
                new_cat("Cat 2::Sub 2", "cat2::sub2", 5),
                new_cat("Cat 3", "cat3", 6),
            ])
            .execute(&mut conn)
            .await
            .unwrap();

        let cats = Category::toplevel(&mut conn, "crates", 2, 0)
            .await
            .unwrap()
            .into_iter()
            .map(|c| (c.category, c.crates_cnt))
            .collect::<Vec<_>>();
        let expected = vec![("Cat 2".to_string(), 12), ("Cat 3".to_string(), 6)];
        assert_eq!(expected, cats);

        let cats = Category::toplevel(&mut conn, "crates", 2, 1)
            .await
            .unwrap()
            .into_iter()
            .map(|c| (c.category, c.crates_cnt))
            .collect::<Vec<_>>();
        let expected = vec![("Cat 3".to_string(), 6), ("Cat 1".to_string(), 3)];
        assert_eq!(expected, cats);
    }

    #[tokio::test]
    async fn category_parent_categories_includes_path_to_node_with_count() {
        use self::categories;

        let new_cat = |category, slug, crates_cnt| {
            (
                categories::category.eq(category),
                categories::slug.eq(slug),
                categories::crates_cnt.eq(crates_cnt),
            )
        };

        let test_db = TestDatabase::new();
        let mut conn = test_db.async_connect().await;

        insert_into(categories::table)
            .values(&vec![
                new_cat("Cat 1", "cat1", 1),
                new_cat("Cat 1::sub1", "cat1::sub1", 2),
                new_cat("Cat 1::sub2", "cat1::sub2", 2),
                new_cat("Cat 1::sub1::subsub1", "cat1::sub1::subsub1", 2),
                new_cat("Cat 2", "cat2", 3),
                new_cat("Cat 2::Sub 1", "cat2::sub1", 4),
                new_cat("Cat 2::Sub 2", "cat2::sub2", 5),
                new_cat("Cat 3", "cat3", 200),
            ])
            .execute(&mut conn)
            .await
            .unwrap();

        let cat: Category = Category::by_slug("cat1::sub1")
            .first(&mut conn)
            .await
            .unwrap();

        let subcats = cat.subcategories(&mut conn).await.unwrap();
        let parents = cat.parent_categories(&mut conn).await.unwrap();

        assert_eq!(parents.len(), 1);
        assert_eq!(parents[0].slug, "cat1");
        assert_eq!(parents[0].crates_cnt, 7);
        assert_eq!(subcats.len(), 1);
        assert_eq!(subcats[0].slug, "cat1::sub1::subsub1");
    }
}
