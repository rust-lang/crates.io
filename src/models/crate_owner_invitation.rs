use chrono::{NaiveDateTime, Utc};
use diesel::prelude::*;
use diesel_async::scoped_futures::ScopedFutureExt;
use diesel_async::{AsyncConnection, AsyncPgConnection, RunQueryDsl};
use secrecy::SecretString;

use crate::config;
use crate::models::{CrateOwner, OwnerKind};
use crate::schema::{crate_owner_invitations, crate_owners, crates};

#[derive(Debug)]
pub enum NewCrateOwnerInvitationOutcome {
    AlreadyExists,
    InviteCreated { plaintext_token: SecretString },
}

#[derive(Clone, Debug, Insertable)]
#[diesel(table_name = crate_owner_invitations, check_for_backend(diesel::pg::Pg))]
pub struct NewCrateOwnerInvitation {
    pub invited_user_id: i32,
    pub invited_by_user_id: i32,
    pub crate_id: i32,
}

impl NewCrateOwnerInvitation {
    pub async fn create(
        &self,
        conn: &mut AsyncPgConnection,
        config: &config::Server,
    ) -> QueryResult<NewCrateOwnerInvitationOutcome> {
        // Before actually creating the invite, check if an expired invitation already exists
        // and delete it from the database. This allows obtaining a new invite if the old one
        // expired, instead of returning "already exists".
        conn.transaction(|conn| {
            async move {
                // This does a SELECT FOR UPDATE + DELETE instead of a DELETE with a WHERE clause to
                // use the model's `is_expired` method, centralizing our expiration checking logic.
                let existing: Option<CrateOwnerInvitation> = crate_owner_invitations::table
                    .find((self.invited_user_id, self.crate_id))
                    .for_update()
                    .first(conn)
                    .await
                    .optional()?;

                if let Some(existing) = existing {
                    if existing.is_expired(config) {
                        diesel::delete(&existing).execute(conn).await?;
                    }
                }
                QueryResult::Ok(())
            }
            .scope_boxed()
        })
        .await?;

        let res: Option<CrateOwnerInvitation> = diesel::insert_into(crate_owner_invitations::table)
            .values(self)
            // The ON CONFLICT DO NOTHING clause results in not creating the invite if another one
            // already exists. This does not cause problems with expired invitation as those are
            // deleted before doing this INSERT.
            .on_conflict_do_nothing()
            .get_result(conn)
            .await
            .optional()?;

        Ok(match res {
            Some(record) => NewCrateOwnerInvitationOutcome::InviteCreated {
                plaintext_token: record.token,
            },
            None => NewCrateOwnerInvitationOutcome::AlreadyExists,
        })
    }
}

/// The model representing a row in the `crate_owner_invitations` database table.
#[derive(Clone, Debug, Identifiable, Queryable)]
#[diesel(primary_key(invited_user_id, crate_id))]
pub struct CrateOwnerInvitation {
    pub invited_user_id: i32,
    pub invited_by_user_id: i32,
    pub crate_id: i32,
    pub created_at: NaiveDateTime,
    #[diesel(deserialize_as = String)]
    pub token: SecretString,
}

impl CrateOwnerInvitation {
    pub async fn find_by_id(
        user_id: i32,
        crate_id: i32,
        conn: &mut AsyncPgConnection,
    ) -> QueryResult<Self> {
        crate_owner_invitations::table
            .find((user_id, crate_id))
            .first::<Self>(conn)
            .await
    }

    pub async fn find_by_token(token: &str, conn: &mut AsyncPgConnection) -> QueryResult<Self> {
        crate_owner_invitations::table
            .filter(crate_owner_invitations::token.eq(token))
            .first::<Self>(conn)
            .await
    }

    pub async fn accept(
        self,
        conn: &mut AsyncPgConnection,
        config: &config::Server,
    ) -> Result<(), AcceptError> {
        use diesel_async::scoped_futures::ScopedFutureExt;
        use diesel_async::{AsyncConnection, RunQueryDsl};

        if self.is_expired(config) {
            let crate_name: String = crates::table
                .find(self.crate_id)
                .select(crates::name)
                .first(conn)
                .await?;

            return Err(AcceptError::Expired { crate_name });
        }

        conn.transaction(|conn| {
            async move {
                diesel::insert_into(crate_owners::table)
                    .values(&CrateOwner {
                        crate_id: self.crate_id,
                        owner_id: self.invited_user_id,
                        created_by: self.invited_by_user_id,
                        owner_kind: OwnerKind::User,
                        email_notifications: true,
                    })
                    .on_conflict(crate_owners::table.primary_key())
                    .do_update()
                    .set(crate_owners::deleted.eq(false))
                    .execute(conn)
                    .await?;

                diesel::delete(&self).execute(conn).await?;

                Ok(())
            }
            .scope_boxed()
        })
        .await
    }

    pub async fn decline(self, conn: &mut AsyncPgConnection) -> QueryResult<()> {
        // The check to prevent declining expired invitations is *explicitly* missing. We do not
        // care if an expired invitation is declined, as that just removes the invitation from the
        // database.

        diesel::delete(&self).execute(conn).await?;
        Ok(())
    }

    pub fn is_expired(&self, config: &config::Server) -> bool {
        self.expires_at(config) <= Utc::now().naive_utc()
    }

    pub fn expires_at(&self, config: &config::Server) -> NaiveDateTime {
        self.created_at + config.ownership_invitations_expiration
    }
}

#[derive(Debug, thiserror::Error)]
pub enum AcceptError {
    #[error(transparent)]
    Diesel(#[from] diesel::result::Error),
    #[error("The invitation has expired")]
    Expired { crate_name: String },
}
