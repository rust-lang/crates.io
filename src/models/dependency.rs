use diesel::sql_types::{BigInt, Text};

use crate::models::{Crate, Version};
use crate::schema::*;
use crates_io_diesel_helpers::pg_enum;
use crates_io_index::DependencyKind as IndexDependencyKind;

#[derive(Identifiable, Associations, Debug, Queryable, QueryableByName, Selectable)]
#[diesel(
    table_name = dependencies,
    check_for_backend(diesel::pg::Pg),
    belongs_to(Version),
    belongs_to(Crate),
)]
pub struct Dependency {
    pub id: i32,
    pub version_id: i32,
    pub crate_id: i32,
    pub req: String,
    pub optional: bool,
    pub default_features: bool,
    pub features: Vec<String>,
    pub target: Option<String>,
    pub kind: DependencyKind,
    pub explicit_name: Option<String>,
}

#[derive(Debug, QueryableByName)]
pub struct ReverseDependency {
    #[diesel(embed)]
    pub dependency: Dependency,
    #[diesel(sql_type = BigInt)]
    pub crate_downloads: i64,
    #[diesel(sql_type = Text, column_name = crate_name)]
    pub name: String,
}

pg_enum! {
    pub enum DependencyKind {
        Normal = 0,
        Build = 1,
        Dev = 2,
    }
}

impl From<IndexDependencyKind> for DependencyKind {
    fn from(dk: IndexDependencyKind) -> Self {
        match dk {
            IndexDependencyKind::Normal => DependencyKind::Normal,
            IndexDependencyKind::Build => DependencyKind::Build,
            IndexDependencyKind::Dev => DependencyKind::Dev,
        }
    }
}

impl From<DependencyKind> for IndexDependencyKind {
    fn from(dk: DependencyKind) -> Self {
        match dk {
            DependencyKind::Normal => IndexDependencyKind::Normal,
            DependencyKind::Build => IndexDependencyKind::Build,
            DependencyKind::Dev => IndexDependencyKind::Dev,
        }
    }
}
