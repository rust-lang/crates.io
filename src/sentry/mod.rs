use crate::config::SentryConfig;
use sentry::protocol::Event;
use sentry::{ClientInitGuard, ClientOptions, TransactionContext};
use std::sync::Arc;

/// Initializes the Sentry SDK from the environment variables.
///
/// If `SENTRY_DSN_API` is not set then Sentry will not be initialized,
/// otherwise it is required to be a valid DSN string. `SENTRY_ENV_API` must
/// be set if a DSN is provided.
///
/// `HEROKU_SLUG_COMMIT`, if present, will be used as the `release` property
/// on all events.
pub fn init() -> Option<ClientInitGuard> {
    let config = match SentryConfig::from_environment() {
        Ok(config) => config,
        Err(error) => {
            warn!(%error, "Failed to read Sentry configuration from environment");
            return None;
        }
    };

    let traces_sampler = move |ctx: &TransactionContext| -> f32 {
        if let Some(sampled) = ctx.sampled() {
            return if sampled { 1.0 } else { 0.0 };
        }

        let op = ctx.operation();
        if op == "http.server" {
            let is_download_endpoint =
                ctx.name().starts_with("GET /api/v1/crates/") && ctx.name().ends_with("/download");

            if is_download_endpoint {
                // Reduce the sample rate for the download endpoint, since we have significantly
                // more traffic on that endpoint compared to the rest
                return config.traces_sample_rate / 100.;
            } else if ctx.name() == "PUT /api/v1/crates/new" {
                // Record all traces for crate publishing
                return 1.;
            } else if ctx.name().starts_with("GET /api/private/metrics/") {
                // Ignore all traces for internal metrics collection
                return 0.;
            }
        } else if op == "swirl.perform" || op == "admin.command" {
            // Record all traces for background tasks and admin commands
            return 1.;
        } else if op == "swirl.run" || op == "server.run" {
            // Ignore top-level span from the background worker and http server
            return 0.;
        }

        config.traces_sample_rate
    };

    let before_send = |mut event: Event<'static>| {
        // Remove cookies from the request to avoid sending sensitive
        // information like the `cargo_session`.
        if let Some(request) = &mut event.request {
            request.cookies.take();
        }

        Some(event)
    };

    let opts = ClientOptions {
        auto_session_tracking: true,
        dsn: config.dsn,
        environment: config.environment.map(Into::into),
        release: config.release.map(Into::into),
        before_send: Some(Arc::new(before_send)),
        session_mode: sentry::SessionMode::Request,
        traces_sampler: Some(Arc::new(traces_sampler)),
        ..Default::default()
    };

    Some(sentry::init(opts))
}
