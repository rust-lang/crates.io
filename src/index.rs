//! This module contains the glue code between our database and the index files
//! and is used by the corresponding background jobs to generate the
//! index files.

use crate::models::{Crate, Dependency, Version};
use crate::schema::crates;
use anyhow::Context;
use crates_io_index::features::split_features;
use diesel::prelude::*;
use diesel_async::{AsyncPgConnection, RunQueryDsl};
use sentry::Level;

#[instrument(skip_all, fields(krate.name = ?name))]
pub async fn get_index_data(
    name: &str,
    conn: &mut AsyncPgConnection,
) -> anyhow::Result<Option<String>> {
    debug!("Looking up crate by name");
    let krate = crates::table
        .select(Crate::as_select())
        .filter(crates::name.eq(name))
        .first::<Crate>(conn)
        .await
        .optional();

    let Some(krate) = krate? else {
        return Ok(None);
    };

    debug!("Gathering remaining index data");
    let crates = index_metadata(&krate, conn)
        .await
        .context("Failed to gather index metadata")?;

    // This can sometimes happen when we delete versions upon owner request
    // but don't realize that the crate is now left with no versions at all.
    //
    // In this case we will delete the crate from the index and log a warning to
    // Sentry to clean this up in the database.
    if crates.is_empty() {
        let message = format!("Crate `{name}` has no versions left");
        sentry::capture_message(&message, Level::Warning);

        return Ok(None);
    }

    debug!("Serializing index data");
    let mut bytes = Vec::new();
    crates_io_index::write_crates(&crates, &mut bytes)
        .context("Failed to serialize index metadata")?;

    let str = String::from_utf8(bytes).context("Failed to decode index metadata as utf8")?;

    Ok(Some(str))
}

/// Gather all the necessary data to write an index metadata file
pub async fn index_metadata(
    krate: &Crate,
    conn: &mut AsyncPgConnection,
) -> QueryResult<Vec<crates_io_index::Crate>> {
    let mut versions: Vec<Version> = Version::belonging_to(krate).load(conn).await?;

    // We sort by `created_at` by default, but since tests run within a
    // single database transaction the versions will all have the same
    // `created_at` timestamp, so we sort by semver as a secondary key.
    versions.sort_by_cached_key(|k| (k.created_at, semver::Version::parse(&k.num).ok()));

    let deps: Vec<(Dependency, String)> = Dependency::belonging_to(&versions)
        .inner_join(crates::table)
        .select((Dependency::as_select(), crates::name))
        .load(conn)
        .await?;

    let deps = deps.grouped_by(&versions);

    versions
        .into_iter()
        .zip(deps)
        .map(|(version, deps)| {
            let mut deps = deps
                .into_iter()
                .map(|(dep, name)| {
                    // If this dependency has an explicit name in `Cargo.toml` that
                    // means that the `name` we have listed is actually the package name
                    // that we're depending on. The `name` listed in the index is the
                    // Cargo.toml-written-name which is what cargo uses for
                    // `--extern foo=...`
                    let (name, package) = match dep.explicit_name {
                        Some(explicit_name) => (explicit_name, Some(name)),
                        None => (name, None),
                    };

                    crates_io_index::Dependency {
                        name,
                        req: dep.req,
                        features: dep.features,
                        optional: dep.optional,
                        default_features: dep.default_features,
                        kind: Some(dep.kind.into()),
                        package,
                        target: dep.target,
                    }
                })
                .collect::<Vec<_>>();

            deps.sort();

            let features = version.features().unwrap_or_default();
            let (features, features2) = split_features(features);

            let (features2, v) = if features2.is_empty() {
                (None, None)
            } else {
                (Some(features2), Some(2))
            };

            let krate = crates_io_index::Crate {
                name: krate.name.clone(),
                vers: version.num.to_string(),
                cksum: version.checksum,
                yanked: Some(version.yanked),
                deps,
                features,
                links: version.links,
                rust_version: version.rust_version,
                features2,
                v,
            };

            Ok(krate)
        })
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::schema::users;
    use crate::tests::builders::{CrateBuilder, VersionBuilder};
    use chrono::{Days, Utc};
    use crates_io_test_db::TestDatabase;
    use diesel_async::AsyncConnection;
    use insta::assert_json_snapshot;

    #[tokio::test]
    async fn test_index_metadata() {
        let test_db = TestDatabase::new();
        let mut conn = test_db.connect();
        let mut async_conn = AsyncPgConnection::establish(test_db.url()).await.unwrap();

        let user_id = diesel::insert_into(users::table)
            .values((
                users::name.eq("user1"),
                users::gh_login.eq("user1"),
                users::gh_id.eq(42),
                users::gh_access_token.eq("some random token"),
            ))
            .returning(users::id)
            .get_result::<i32>(&mut async_conn)
            .await
            .unwrap();

        let created_at_1 = Utc::now()
            .checked_sub_days(Days::new(14))
            .unwrap()
            .naive_utc();

        let created_at_2 = Utc::now()
            .checked_sub_days(Days::new(7))
            .unwrap()
            .naive_utc();

        let fooo = CrateBuilder::new("foo", user_id)
            .version(VersionBuilder::new("0.1.0"))
            .expect_build(&mut conn);

        let metadata = index_metadata(&fooo, &mut async_conn).await.unwrap();
        assert_json_snapshot!(metadata);

        let bar = CrateBuilder::new("bar", user_id)
            .version(
                VersionBuilder::new("1.0.0-beta.1")
                    .created_at(created_at_1)
                    .yanked(true),
            )
            .version(VersionBuilder::new("1.0.0").created_at(created_at_1))
            .version(
                VersionBuilder::new("2.0.0")
                    .created_at(created_at_2)
                    .dependency(&fooo, None),
            )
            .version(VersionBuilder::new("1.0.1").checksum("0123456789abcdef"))
            .expect_build(&mut conn);

        let metadata = index_metadata(&bar, &mut async_conn).await.unwrap();
        assert_json_snapshot!(metadata);
    }
}
