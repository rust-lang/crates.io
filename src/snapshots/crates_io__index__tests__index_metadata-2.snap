---
source: src/index.rs
expression: metadata
---
[
  {
    "name": "bar",
    "vers": "1.0.0-beta.1",
    "deps": [],
    "cksum": "                                                                ",
    "features": {},
    "yanked": true
  },
  {
    "name": "bar",
    "vers": "1.0.0",
    "deps": [],
    "cksum": "                                                                ",
    "features": {},
    "yanked": false
  },
  {
    "name": "bar",
    "vers": "2.0.0",
    "deps": [
      {
        "name": "foo",
        "req": ">= 0",
        "features": [],
        "optional": false,
        "default_features": false,
        "target": null,
        "kind": "normal"
      }
    ],
    "cksum": "                                                                ",
    "features": {},
    "yanked": false
  },
  {
    "name": "bar",
    "vers": "1.0.1",
    "deps": [],
    "cksum": "0123456789abcdef                                                ",
    "features": {},
    "yanked": false
  }
]
